File: GameServer/pyserver/lib/transport.py
import zmq
import json
import traceback
import time
import threading
import logging
from userOverrides import Instance,Protocol
from res import actions
import config

class Transport():
    data_push = []
    data_sub = []
    #response = {}
    # this will have all the game instances
    instances = []

    def __init__(self):
        self.threads = [
            threading.Thread(target=self.push_server),
            threading.Thread(target=self.alive)
        ]
        # step 1 start all servers
        for thread in self.threads:
            thread.start()

        # step 2 connect to api server
        self.send(action=actions.connect, api=True)

        # step 3 subscribe to client shared key
        self.subscribe(config.SHARED_KEY)

    # call this method to subscribe to a new key
    def subscribe(self, key):
        thread = threading.Thread(target=self.subscribe_server, args=(key,))
        thread.start()
        self.threads.append(thread)

    # call this to push to the api server
    def send(self, query={}, action=None, api=False, shared=False):
        query['time'] = time.time()
        if action:
            query['action'] = action
        if api:
            query['api_key'] = config.API_KEY
        if shared:
            query['shared_key'] = config.SHARED_KEY

        self.data_push.append(json.dumps(query))


    def parse_response(self, data):
        _action = data['action']
        if _action == actions.get_games:
            data['client_shared_key'] = data['shared_key']
            data['games'] = Protocol.get_games(self, data['client_shared_key'])
            self.send(data, action=actions.game_list, shared=True)

        if _action == actions.new_game:
            new_instance=Instance(data['session_key'])
            new_instance = Protocol.new_game(new_instance,data['shared_key'])
            self.instances.append(new_instance)

        if _action == actions.select_game:
            instance = self.get_instance(data['session_key'])
            instance.join_instance(data['shared_key'])

        if _action == actions.event:
            instance = self.get_instance(data['session_key'])
            # set game state to processed game state
            # todo make game_state modification lock and process type
            instance.game_state = Protocol.event(data)
            data=instance.game_state.to_json()
            data['session_key']=instance.session_key
            self.send(data,action=actions.game_state,shared=True)

        if _action == actions.dead:
            instance = self.get_instance(data['session_key'])
            instance.leave_instance(data['shared_key'])
            Protocol.dead(data)
            if instance.status == 'removed':
                del instance

        if _action == actions.disconnect:
            instance = self.get_instance(data['session_key'])
            instance.leave_instance(data['shared_key'])
            Protocol.disconnect(data)
            if instance.status == 'removed':
                del instance

    def get_instance(self,session_key):
        for instance in self.instances:
            if str(instance.session_key) == str(session_key):
                return instance

    # this keeps sending packet to server to tell that it is alive
    def alive(self):
        while True:
            data={}
            self.send(data,action=actions.alive, shared=True)
            time.sleep(config.ALIVE_PULSE)

    def push_server(self):
        context = zmq.Context()
        socket = context.socket(zmq.PUSH)

        for port in config.PORT_PUSH:
            socket.connect("tcp://" + config.HOST + ":" + port)

        while True:
            try:
                for data in self.data_push:
                    print logging.debug(data, "[PUSH]")
                    socket.send(data)
                    self.data_push.remove(data)

                # to decrease CPU usage
                time.sleep(0.001)
            except:
                traceback.print_exc()


    # DONT start this right in the beginning coz u dont know what is the session key of the game instance
    # todo see whether client has to subscribe to many session keys
    def subscribe_server(self, key):
        context = zmq.Context()
        socket = context.socket(zmq.SUB)
        socket.connect("tcp://" + config.HOST + ":" + config.PORT_SUB)

        socket.setsockopt(zmq.SUBSCRIBE, key)
        while True:
            try:
                data = socket.recv()
                data = " ".join(data.split(" ")[1:])
                print logging.debug(data, "[SUBSCRIBE]")
                #self.data_sub.append(json.loads(data))
                thread = threading.Thread(target=self.parse_response, args=(json.loads(data),))
                thread.start()

                # to decrease CPU usage
                time.sleep(0.001)
            except:
                traceback.print_exc()

File: GameServer/cpp/transport.cpp
#include<pthread.h>
#include<time.h>
#include<sstream>
#define len(x) (sizeof(x)/sizeof(x[0]))

using namespace std;
// use usleep instead of sleep
class Transport{
	public :
	vector<pthread_t> threads;
	vector<string> data_push, data_sub;
	vector<Instance> instances;
	
	Transport(void(*run)()){
		thread push_server_thread(push_server,this);
		thread alive_thread(alive,this);
		thread subscribe_server_thread(subscribe_server,this,SHARED_KEY);
		thread run_thread(run);

		send(Json::Value::null,ACTION_CONNECT,true,false);

		alive_thread.join();
		push_server_thread.join();
		subscribe_server_thread.join();
		run_thread.join();
	}
	void subscribe(string);
	void send(Json::Value, string, bool, bool);
	static void parse_response(Transport*,Json::Value);
	Instance* get_instance(string);
	static void alive(Transport*);
	static void push_server(Transport*);
	static void subscribe_server(Transport*,string);
};

void Transport::subscribe(string key){
	thread subscribe_server_thread(subscribe_server,this,key);
	subscribe_server_thread.join();
}
void Transport::send(Json::Value query, string action="", bool api=false, bool shared=false){
	time_t t = time(0);
	query["time"]= (int) t;

	if(action!="")
		query["action"] = action;
	if(api)
		query["api_key"] = API_KEY; 
	if(shared)
		query["shared_key"] = SHARED_KEY;
	logging::debug(query.toStyledString());	
	data_push.push_back(query.toStyledString());
}
void Transport::parse_response(Transport* self, Json::Value data){
	string action = data["action"].asString();
	if (action == ACTION_GET_GAMES){
		data["client_shared_key"] = data["shared_key"];
		data["games"] = Protocol::get_games(self->instances);
		self->send(data, ACTION_GAME_LIST, false, true);
	}
	else if (action == ACTION_NEW_GAME){
		Instance new_instance = Instance(data["session_key"].asString());
		new_instance = Protocol::new_game(new_instance,data["shared_key"].asString());
		self->instances.push_back(new_instance);
	}
	else if(action == ACTION_SELECT_GAME){
		Instance* instance = self->get_instance(data["session_key"].asString());
		instance->join_instance(data["shared_key"].asString());
	}
	else if(action == ACTION_EVENT){
		Instance* instance = self->get_instance(data["session_key"].asString());
		instance->game_state = Protocol::event(data);
		Json::Value data = instance->game_state.data;
		data["session_key"] = instance->session_key;
		self->send(data,ACTION_GAME_STATE,false,true);
	}
	else if(action == ACTION_DEAD){
		Instance* instance = self->get_instance(data["session_key"].asString());
		instance->leave_instance(data["shared_key"].asString());
		Protocol::dead(data);
		if(instance->status == "removed")
			free(instance);
	}
	else if(action == ACTION_DISCONNECT){
		Instance* instance = self->get_instance(data["session_key"].asString());
		instance->leave_instance(data["shared_key"].asString());
		Protocol::disconnect(data);
		if(instance->status == "removed")
			free(instance);
	}
}
Instance* Transport::get_instance(string session_key){
	for(vector<Instance>::iterator it=instances.begin(); it!=instances.end(); ++it){
		if((*it).session_key==session_key)
			return &(*it);
	}
}
int i=0;
void Transport::alive(Transport* T){
	while(true){
		T->send(Json::Value::null,ACTION_ALIVE,false,true);
		chrono::milliseconds dura(1000*ALIVE_PULSE);
		this_thread::sleep_for(dura);
	}
}
void Transport::push_server(Transport* self){
	zmq::context_t context (1);
    	zmq::socket_t socket (context, ZMQ_PUSH);
	for(int i=0;i<len(PORT_PUSH);i++){
		string addr = "tcp://" + HOST + ":" + PORT_PUSH[i]; 
    		socket.connect(addr.c_str());

	}
	while(true){
		vector<string> data_copy = self->data_push;
		self->data_push.clear();
		for(int i=0;i<data_copy.size();i++){
			char *data = (char*)(data_copy[i].c_str());
			zmq::message_t push_message(strlen(data))	;
			memcpy ((void *) push_message.data(), data, strlen(data));
        		socket.send (push_message);
		}
		usleep(10);
	}
}
void Transport::subscribe_server(Transport* self, string key){
	cout<<key;
	zmq::context_t context (1);
    	zmq::socket_t socket (context, ZMQ_SUB);
	string addr = "tcp://" + HOST + ":" + PORT_SUB; 
	socket.connect(addr.c_str());
	socket.setsockopt(ZMQ_SUBSCRIBE,key.c_str(),strlen(key.c_str()));
	while(true){
		zmq::message_t reply_data;
		socket.recv(&reply_data);
		stringstream ss;
		string data;
		ss << (char*)reply_data.data();
		data = ss.str();
		data = data.substr(data.find("{"));
		data = data.substr(0,data.find_last_of("}")+1);
		
		logging::debug(data.c_str(), "[SUBSCRIBE]");
		thread parse_response_thread(parse_response,self,JSON::parse(data.c_str()));
		parse_response_thread.join();
		usleep(10);
	}
}
